teps/0013-graceful-sidecar-termination.md

```diff
@@ -0,0 +1,226 @@
+
+## Design Details
+
+A `Task` can have multiple sidecar defined each behaving according to the flag. Injected sidecars are considered as `waitForsidecar` false since that's the
+default behavior.
+
+Example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+spec:
+  sidecars:
+  - env:
+    - name: POD_NAME
+      valueFrom:
+        fieldRef:
+          fieldPath: metadata.name
+    - name: E2E_CONTAINER_NAME
+      value: step-e2e
+    - name: NAMESPACE
+      value: test
+    - name: BUILD_NUMBER
+      value: "25"
+    waitForSidecar: true
+    image: rifatshahnewaz/e2e-uploader:prod
+    name: publish-e2e-report
+  - env:
+    - name: POD_NAME
+      valueFrom:
+        fieldRef:
+          fieldPath: metadata.name
+    - name: LOG_STREAMER_CONTAINER_NAME
+      value: step-log-streamer
+    - name: NAMESPACE
+      value: test
+    - name: BUILD_NUMBER
+      value: "25"
+    waitForSidecar: false
+    image: rifatshahnewaz/log-streamer:prod
+    name: log-streamer
+
+    volumeMounts:
+    - mountPath: /coverage-results
+      name: e2e-report
+```
+
+Golang Example:
+```
+s.TektonTask.Spec.Sidecars = []pipelinev1beta1.Sidecar{
+	{
+		Container: corev1.Container{
+			Name:            "publish-e2e-report",
+			Image:           "e2e-runner-image",
+			Env: []corev1.EnvVar{
+				{
+					Name: "POD_NAME",
+					ValueFrom: &corev1.EnvVarSource{
+						FieldRef: &corev1.ObjectFieldSelector{
+							FieldPath: "metadata.name",
+						},
+					},
+				},
+				{Name: "E2E_CONTAINER_NAME", Value: "step-e2e"},
+				{Name: "NAMESPACE", Value: yaml.ProductName},
+				{Name: "BUILD_NUMBER", Value: buildNumber},
+			},
+			VolumeMounts: []corev1.VolumeMount{{
+				Name:      "e2e-volume",
+				ReadOnly:  false,
+				MountPath: "/coverage-results",
+			}},
+		},
+		waitForSidecar: true,	
+	},
+		{
+		Container: corev1.Container{
+			Name:            "log-streamer",
+			Image:           "log-streamer-image",
+			Env: []corev1.EnvVar{
+				{
+					Name: "POD_NAME",
+					ValueFrom: &corev1.EnvVarSource{
+						FieldRef: &corev1.ObjectFieldSelector{
+							FieldPath: "metadata.name",
+						},
+					},
+				},
+				{Name: "LOG_STREAMER_CONTAINER_NAME", Value: "log-streamer"},
+				{Name: "NAMESPACE", Value: yaml.ProductName},
+				{Name: "BUILD_NUMBER", Value: buildNumber},
+			},
+		},
+	},
+}
+
+```
+
+What happens when the taskrun has reached a timeout but the sidecar is still running?
+- The sidecar should be terminated using a Nop Image in such case. 
+
+## Advantages
+
+* No need to introduce another step for simpler tasks.
+* Support Task behaviours not yet provided for by Tekton: cleanup, uploading, report generation after Steps have succeeded or failed.
+
+## Test Plan
+
+* e2e and unit tests
+
+## Drawbacks
+Discussed in Cons.
+
+**Pros/Cons:**
+Pros:
+Discussed above.
+
+Cons:
+Can't think of any. Unless the pipeline author adds a sidecar with this flag enabled and the sidecars' cmd is an infinite one. Causing the pipeline to hang
```

Comment:
I think this is a good drawback to call out.  Suggest formatting it something like:

```
## Drawbacks

* When the waitForSidecar flag is used and a timeout of 0 is specified for the TaskRun, a Sidecar that fails to terminate could result in a TaskRun or PipelineRun executing forever until a user manually intervenes to stop the Pod.
```